### DISCUSSION TITLE
Sound distorted when begin writing screen

### DISCUSSION THREAD

**Discussion by mignon-ne** [moononournation/Arduino_GFX#463] [Q&A]
HW: ESP32-4827S043N (ESP32-S3-WROOM-1, 240MHz, 512KB SRAM, 384KB ROM, 8M PSRAM, 16MB Flash)
Display resolution is 480*272
DAC: MAX98357A

Audio Library: ESP32-audioI2S (schreibfaul1)
Graphic: GFX_Library_for_Arduino (1.4.6)

As the board limitation, I have to replace pin that connect to SD Card reader to MAX98357
#define I2S_DOUT		11
#define I2S_BCLK		12
#define I2S_LRC			13

My testing code is from the simple example from GFX and Audio.
The audio sound work fine and last for very long without any problem.
But when begin to esecute any writing (gfx->println) for few times the audio sound distorted and very low, at some point not hearing anything.
There is no special event in audio_info.

Try audio task pin to core has the same symptom.

Any idea or suggestion is appreciated.
https://github.com/moononournation/Arduino_GFX/assets/16378483/2486abc5-0bef-4e99-ae5d-be338bf40156

[VDO](https://youtu.be/83h0qq9o888)
```
/*******************************************************************************
 * Start of Arduino_GFX setting
 * 
 * Arduino_GFX try to find the settings depends on selected board in Arduino IDE
 * Or you can define the display dev kit not in the board list
 * Defalult pin list for non display dev kit:

 ******************************************************************************/
#include <Arduino_GFX_Library.h>
#include <Audio.h>
//#include <FS.h>
//#include <SPIFFS.h>

#define GFX_BL 2 //DF_GFX_BL // default backlight pin, you may replace DF_GFX_BL to actual backlight pin
/* More dev device declaration: https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration */
#if defined(DISPLAY_DEV_KIT)
Arduino_GFX *gfx = create_default_Arduino_GFX();
#else /* !defined(DISPLAY_DEV_KIT) */

/* More data bus class: https://github.com/moononournation/Arduino_GFX/wiki/Data-Bus-Class */
//Arduino_DataBus *bus = create_default_Arduino_DataBus();

/* More display class: https://github.com/moononournation/Arduino_GFX/wiki/Display-Class */
//Arduino_GFX *gfx = new Arduino_ILI9341(bus, DF_GFX_RST, 0 /* rotation */, false /* IPS */);


Arduino_ESP32RGBPanel *bus = new Arduino_ESP32RGBPanel(
	40 /* DE */, 41 /* VSYNC */, 39 /* HSYNC */, 42 /* PCLK */,
	45 /* R0 */, 48 /* R1 */, 47 /* R2 */, 21 /* R3 */, 14 /* R4 */,
	5 /* G0 */, 6 /* G1 */, 7 /* G2 */, 15 /* G3 */, 16 /* G4 */, 4 /* G5 */,
	8 /* B0 */, 3 /* B1 */, 46 /* B2 */, 9 /* B3 */, 1 /* B4 */,
	0 /* hsync_polarity */, 1 /* hsync_front_porch */, 1 /* hsync_pulse_width */, 43 /* hsync_back_porch */,
	0 /* vsync_polarity */, 3 /* vsync_front_porch */, 1 /* vsync_pulse_width */, 12 /* vsync_back_porch */,
	1 /* pclk_active_neg */, 10000000 /* prefer_speed */);

Arduino_RGB_Display *gfx = new Arduino_RGB_Display(480 /* width */, 272 /* height */, bus);




#endif /* !defined(DISPLAY_DEV_KIT) */
/*******************************************************************************
 * End of Arduino_GFX setting
 ******************************************************************************/


// I2S Connections ESP32-4827S043, replace SD card
#define I2S_DOUT      11	//22
#define I2S_BCLK      12	//26
#define I2S_LRC       13 //19 USB, 11 is flash

#define FORMAT_SPIFFS_IF_FAILED false //true

#define PRF				Serial.printf
#define PRN				Serial.println
#define PRR				Serial.print


 // Create Audio object
Audio audio;
uint32_t Vol;

const TickType_t xDelay1ms = pdMS_TO_TICKS(1);


void Task_Audio(void *pvParameters) // This is a task.
{
	while(true)  {
		audio.loop();
		delay(1);	// required
		//vTaskDelay(xDelay1ms);
	};
}

void audio_info(const char *info){
    Serial.print("info        "); Serial.println(info);
}
void audio_id3data(const char *info){  //id3 metadata
    Serial.print("id3data     ");Serial.println(info);
}

void audio_showstation(const char *info){
    Serial.print("station     ");Serial.println(info);
}
void audio_showstreamtitle(const char *info){
    Serial.print("streamtitle ");Serial.println(info);
}
String ssid =     "xxx_2.4G";
String password = "xxxxxxxx";

void setup(void)
{
	Serial.begin(115200);
	delay(200);

	WiFi.disconnect();
	WiFi.mode(WIFI_STA);
	WiFi.begin(ssid.c_str(), password.c_str());
	while (WiFi.status() != WL_CONNECTED) delay(1500);

	audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT);
	//xTaskCreatePinnedToCore(Task_Audio, "Task_Audio", 10240, NULL, 3, NULL, 0);

	Vol = 8;
	// Set Volume101
	audio.setVolume(Vol); // 0...21
	audio.forceMono(true);


	gfx->begin();
	//gfx->setRotation(0);
	//gfx->fillScreen(BLACK);

	#ifdef GFX_BL
	pinMode(GFX_BL, OUTPUT);
	digitalWrite(GFX_BL, HIGH);
	#endif
	
	delay(1000); // 5 seconds
	//gfx->setCursor(10, 10);
	//gfx->setTextColor(RED);
	//gfx->println("Hello World!");

	
	//audio.connecttohost("https://github.com/schreibfaul1/ESP32-audioI2S/raw/master/additional_info/Testfiles/Pink-Panther.wav");        // wav
	//audio.connecttohost("http://www.wdr.de/wdrlive/media/einslive.m3u");
	audio.connecttohost("http://d141.rndfnk.com/ard/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3&sid=2ed4mvlrDKOZXXMmk7bsdpKAtMK&token=L0aREcM2_8bC-nWHIXZciRYjY-ReV5GShw_Kcxtbbes&tvf=lBouvdQbwxdkMTQxLnJuZGZuay5jb20");
}


void loop()
{
char in;
static char chin[64];
static int v, i = 0;

	audio.loop();

	while(Serial.available())   {
		in = Serial.read();
		if(in!=0x0A)	{
			chin[i++] = in;
		}else	{
			chin[i] = 0;
			//Serial.println(chin);
			if(chin[0] == 'd')	{
				gfx->setCursor(random(gfx->width()), random(gfx->height()));
				gfx->setTextColor(random(0xffff), random(0xffff));
				gfx->setTextSize(random(6) , random(6) , random(2));
				gfx->println("Hello World!");
			}else if(chin[0] == 'v')	{
				v = atoi(chin+1);
				if(v <= 21 ) {
					Vol = v;
					PRF("Update volume to %d\n", Vol);
					audio.setVolume(Vol);
				}else
					PRN("Too high, no volume update");
			}
			i = 0;
		}
	};

	delay(1);
}
```

----------------------------------------------------------------------------
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x44c
load:0x403c9700,len:0xbd8
load:0x403cc700,len:0x2a80
entry 0x403c98d0
info        Connect to new host: "http://d141.rndfnk.com/ard/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3&sid=2ed4mvlrDKOZXXMmk7bsdpKAtMK&token=L0aREcM2_8bC-nWHIXZciRYjY-ReV5GShw_Kcxtbbes&tvf=lBouvdQbwxdkMTQxLnJuZGZuay5jb20"
info        PSRAM found, inputBufferSize: 638965 bytes
info        buffers freed, free Heap: 247800 bytes
info        Connection has been established in 153 ms, free Heap: 246696 bytes
info        redirect to new host "http://dispatcher.rndfnk.com/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3"
info        Connect to new host: "http://dispatcher.rndfnk.com/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3"
info        buffers freed, free Heap: 247264 bytes
info        Connection has been established in 163 ms, free Heap: 246820 bytes
info        redirect to new host "http://f131.rndfnk.com/ard/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3&sid=2efk35ixqpj1gkHXsJweMPRbbLr&token=OXd5NhcrOoe5IEUdHl4zPBRjL4nvUMx9DyMN1Iv62Tg&tvf=vZ0cYfxlwxdmMTMxLnJuZGZuay5jb20"
info        Connect to new host: "http://f131.rndfnk.com/ard/wdr/1live/live/mp3/128/stream.mp3?aggregator=app&cid=01FBRZTS1K1TCD4KA2YZ1ND8X3&sid=2efk35ixqpj1gkHXsJweMPRbbLr&token=OXd5NhcrOoe5IEUdHl4zPBRjL4nvUMx9DyMN1Iv62Tg&tvf=vZ0cYfxlwxdmMTMxLnJuZGZuay5jb20"
info        buffers freed, free Heap: 246676 bytes
info        Connection has been established in 101 ms, free Heap: 246232 bytes
station     1Live, Westdeutscher Rundfunk Koeln
info        MP3Decoder has been initialized, free Heap: 245244 bytes , free stack 5292 DWORDs
info        stream ready
info        syncword found at pos 0
info        MP3 decode error -2 : MAINDATA_UNDERFLOW
info        syncword found at pos 78
info        syncword found at pos 0
info        MP3 decode error -6 : INVALID_FRAMEHEADER
info        syncword found at pos 304
info        syncword found at pos 0
info        Channels: 2
info        SampleRate: 48000
info        BitsPerSample: 16
info        BitRate: 128000
info        StreamTitle='Taylor Swift - Cruel Summer'
streamtitle Taylor Swift - Cruel Summer

**Comment by moononournation** [comment]
In most case, increase the I2S buffer

**Reply by mignon-ne** [reply]
Thank you very much for your fast response.
Understand that it is I2S command, but tried dirty test as below.
audio.setBufsize(16*1024, 128*1024);
Change m_maxBlockSize in Audio.h fro 1600->16384
size_t   m_maxBlockSize     = 16384; //1600;

Not help, probably I did it wrong.

**Reply by moononournation** [reply]
I direct change the i2s_config.dma_buf_count and .dma_buf_len before calling i2s_driver_install().

**Reply by mignon-ne** [reply]
Change the buf as below, not help.
    m_i2s_config.dma_buf_count        = 64;	//16;
    m_i2s_config.dma_buf_len          = 256;//512;
But when I Re-fillScreen - the sound quality is back to normal (as shown in VDO).
```
if(chin[0] == 'd')	{
	gfx->setCursor(random(gfx->width()), random(gfx->height()));
	gfx->setTextColor(random(0xffff), random(0xffff));
	gfx->setTextSize(random(6) , random(6) , random(2));
	gfx->println("Hello World!");
}else if(chin[0] == 'b')	{
	gfx->fillScreen(BLACK);
	}else..
```

Wonder, as I understanding the GFX graphic is static (not keep track of what we have on screen) the more we write or update should not impact the memory usages and fill in any color is not clear the memory (I probably wrong).
Any clue that I should look in to, please advice.

https://github.com/moononournation/Arduino_GFX/assets/16378483/03dd3731-9212-474a-8d84-638218d7fc60

.

**Reply by moononournation** [reply]
use playing 25 fps video example, you need to have at least 0.04 seconds audio buffer.
if sample rate is 44100, then: 44100 x 16 bit x 2 channels x 0.04 ~ 8k bytes

**Reply by moononournation** [reply]
and you need to fill it up before darwing

**Reply by mignon-ne** [reply]
Thank you very much, will try.
